pkg/redirectpolicy/manager.go

```diff
@@ -0,0 +1,531 @@
+		}
+	}
+
+	return true, nil
+}
+
+// Event handlers
+
+// AddRedirectPolicy parses the given local redirect policy config, and updates
+// internal state with the config fields.
+func (rpm *RedirectPolicyManager) AddRedirectPolicy(config LRPConfig, svcCache *k8s.ServiceCache) (bool, error) {
+	rpm.Lock()
+	defer rpm.Unlock()
+
+	_, ok := rpm.policyConfigs[config.getPolicyID()]
+	if ok {
+		// TODO Existing policy update
+		return true, nil
+	}
+
+	valid, err := rpm.isValidConfig(config)
+	if !valid {
+		return false, err
+	}
+
+	log.Infof("Add redirect policy %s", config.getPolicyID())
+	// New redirect policy
+	rpm.storePolicyConfig(config)
+
+	if config.serviceID != nil {
+		rpm.services[config.serviceID.String()] = config.getPolicyID()
+		svcAddr := rpm.getSvcFrontend(config, svcCache)
+		if svcAddr != nil && svcAddr.Protocol == lb.L4Type(config.svcPortProto.Protocol) {
+			rpm.updatePolicyConfigFrontend(&config, svcAddr, true)
+		} else {
+			// No service entry yet, create a pending request.
+			rpm.pendingEvents.forService[config.serviceID.String()] = &config
+		}
+	}
+	pods := rpm.getLocalPodsForPolicy(&config, rpm.epManager)
+	if len(pods) == 0 {
+		return true, nil
+	}
+	backends := rpm.getPodBackendsForPolicy(&config, pods...)
+	if len(backends) > 0 {
+		rpm.updatePolicyConfigBackends(&config, true, backends...)
+	}
+
+	return true, nil
+}
+
+// DeleteRedirectPolicy deletes the given local redirect policy config, and deletes
+// the internal state associated with the policy.
+func (rpm *RedirectPolicyManager) DeleteRedirectPolicy(config LRPConfig) (bool, error) {
+	rpm.Lock()
+	defer rpm.Unlock()
+
+	storedConfig := rpm.policyConfigs[config.getPolicyID()]
+	if storedConfig == nil {
+		return false, fmt.Errorf("Local redirect policy to be deleted not found")
+	}
+	if storedConfig.serviceID != nil {
+		rpm.deleteService(*storedConfig.serviceID)
+		delete(rpm.pendingEvents.forService, storedConfig.serviceID.String())
+	} else {
+		rpm.notifyPolicyFrontendDelete(storedConfig)
+	}
+	delete(rpm.policyConfigs, storedConfig.getPolicyID())
+	return false, nil
+}
+
+// OnAddService handles Kubernetes service updates, and updates the internal state
+// for the policy config that might be associated with the service.
+func (rpm *RedirectPolicyManager) OnAddService(svcID k8s.ServiceID, svcCache *k8s.ServiceCache) {
+	rpm.Lock()
+	defer rpm.Unlock()
+	if len(rpm.policyConfigs) == 0 {
+		return
+	}
+
+	// Check if there are any pending events for this service.
+	if config, ok := rpm.pendingEvents.forService[svcID.String()]; ok {
+		svcAddr := rpm.getSvcFrontend(*config, svcCache)
+		if svcAddr != nil {
+			delete(rpm.pendingEvents.forService, svcID.String())
+			rpm.updatePolicyConfigFrontend(config, svcAddr, true)
+		}
+	}
+}
+
+// deleteService deletes config associated with the specified service.
+// Note: Needs to be called with redirect policy manager lock.
+func (rpm *RedirectPolicyManager) deleteService(svcID k8s.ServiceID) {
```

Comment:
nit: We often name such functions like `deleteServiceLocked()` to indicate that a lock must be held. This has the added benefit of showing up in gops stack traces in live environments so is super handy if we somehow get into a weird locking situation.